// @graphql-codegen/visitor-plugin-common/dist/esnext/base-visitor.js
import * as autoBind from 'auto-bind';
import { DEFAULT_SCALARS, normalizeScalars } from './scalars';
import { convertFactory } from './naming';
export class BaseVisitor {
    constructor(rawConfig, additionalConfig, defaultScalars = DEFAULT_SCALARS) {
        this._declarationBlockConfig = {};
        this._parsedConfig = {
            scalars: normalizeScalars({
                ...(defaultScalars || DEFAULT_SCALARS),
                ...(rawConfig.scalars || {}),
            }),
            convert: convertFactory(rawConfig),
            typesPrefix: rawConfig.typesPrefix || '',
            namespacedImportName: rawConfig.namespacedImportName || null,
            externalFragments: rawConfig.externalFragments || [],
            addTypename: !rawConfig.skipTypename,
            nonOptionalTypename: !!rawConfig.nonOptionalTypename,
            ...(additionalConfig || {}),
        };
        autoBind(this);
    }
    get config() {
        return this._parsedConfig;
    }
    get scalars() {
        return this.config.scalars;
    }
    convertName(node, options) {
        const useTypesPrefix = typeof (options && options.useTypesPrefix) === 'boolean' ? options.useTypesPrefix : true;
        return (useTypesPrefix ? this.config.typesPrefix : '') + this.config.convert(node, options);
    }
}
